src/index.ts

```diff
@@ -13,49 +13,100 @@
+
 import {Config, defaultConfig} from './config';
 import {Profiler, ProfilerConfig} from './profiler';
 
 const common: Common = require('@google-cloud/common');
+const metadataAPI = 'http://metadata.google.internal/computeMetadata/v1/';
+
+// Returns true if running on GCE and false otherwise.
+// It is assumed a program is running on GCE if the metadata API can be
+// accessed.
+function onGCE(): Promise<boolean> {
+  return new Promise<boolean>((resolve) => {
+    request.get(
+        {
+          url: metadataAPI,
+          timeout: 2000,
+          headers: {'Metadata-Flavor': 'Google'}
+        },
+        function(err: Error, response) {
+          err ? resolve(false) : resolve(true);
+        });
+  });
+}
+
+// Returns value of metadata field.
+// Throws error if there is a problem accessing metadata API.
+function getField(field: string): Promise<string> {
+  return new Promise<string>((resolve, reject) => {
+    request.get(
+        {url: metadataAPI + field, headers: {'Metadata-Flavor': 'Google'}},
+        function(err: Error, response: request.RequestResponse) {
+          err ? reject(err) : resolve(response.body);
+        });
+  });
+}
 
 // initConfig sets unset values in the configuration to the value retrieved from
-// environment variables, metadata, or the default value specified in
+// environment variables, metadata, or the default values specified in
 // defaultConfig.
-// Returns rejected promise if value that must be set cannot be initialized.
-async function initConfig(config: Config): Promise<ProfilerConfig> {
+// Throws error if value that must be set cannot be initialized.
+// Exported for testing purposes.
+export async function initConfig(config: Config): Promise<ProfilerConfig> {
   config = common.util.normalizeArguments(null, config);
 
-  const envConfig = {
-    logLevel: process.env.GCLOUD_PROFILER_LOGLEVEL,
+  const envConfig: any = {
     projectId: process.env.GCLOUD_PROJECT,
     serviceContext: {
       service: process.env.GAE_SERVICE,
       version: process.env.GAE_VERSION,
     }
   };
 
+  if (process.env.GCLOUD_PROFILER_LOGLEVEL !== undefined) {
+    let envLogLevel = parseInt(process.env.GCLOUD_PROFILER_LOGLEVEL || '', 10);
+    if (envLogLevel !== NaN) {
+      envConfig.logLevel = envLogLevel;
+    }
+  }
+
   let envSetConfig: Config = {};
   if (process.env.hasOwnProperty('GCLOUD_PROFILER_CONFIG')) {
     envSetConfig =
         require(path.resolve(process.env.GCLOUD_PROFILER_CONFIG)) as Config;
   }
 
-  let normalizedConfig = extend(true, {}, defaultConfig, envConfig, config);
+  let normalizedConfig =
+      extend(true, {}, defaultConfig, envSetConfig, envConfig, config);
 
-  if (normalizedConfig.serviceContext.service === undefined) {
-    throw new Error('service name must be specified in the configuration');
+  // fetch instance and zone from metadata if not specified and on GCE
+  if (await onGCE()) {
```

Comment:
Reverse the order of these ifs. If you have a locally provided instance/zone, you don't need to await `onGCE` (i.e. no need to add a network latency when you don't need it). 

Secondly, I don't think you really need to check if you're on GCE before attempting to fetch instance/zone. You should fetch these values (if not provided by user) from the metadata service blindly, and if you're not on GCE the requests will fail.

Thirdly, you should fetch instance and zone in parallel rather than serializing those latencies. Something like:

```
const [instance, zone] = await Promsie.all([getField('instance/name'), getField('instance/zone')]);
```

Reply:
I generally agree with your comment, but I am a little concerned about changing this to what you've suggested, because I believe we want to require that the instance and zone are specified when code is being run from GCE, but not require that these fields are specified off of GCE. 

I'm currently echoing the pattern used by the go profiling agent (https://code.googlesource.com/gocloud/+/master/profiler/profiler.go#363) 

@aalexand -- thoughts?

Reply:
> we want to require that the instance and zone are specified when code is being run from GCE, but not require that these fields are specified off of GCE.

My suggestion doesn't contradict this. I am suggesting that that the 'onGCE' check is not necessary and you get the exact same outcome regardless.

Reply:
I was worried about the situation in which instance or zone might not be set in the metadata.
But I think this should be okay, given that it does look like these fields are set in metadata.